File: EmarsysMobileShop/EmarsysMobileShop/ItemsManager.m
//
//  ItemsManager.m
//  EmarsysMobileShop
//

#import "ItemsManager.h"

@implementation ItemsManager

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    return self.items.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    UITableViewCell *cell =
        [tableView dequeueReusableCellWithIdentifier:@"ItemCell"];
    NSAssert(cell, @"No cell");

    Item *item = self.items[indexPath.row];

    [(UILabel *)[cell.contentView viewWithTag:11] setText:item.title];
    [(UILabel *)[cell.contentView viewWithTag:12]
        setText:[NSString stringWithFormat:@"$%.02f", item.price]];

    NSURL *url = [NSURL URLWithString:item.image];

    NSURLSessionTask *task = [[NSURLSession sharedSession]
          dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data,
                            NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {

          if (!data) {
              return;
          }

          UIImage *image = [UIImage imageWithData:data];
          if (!image) {
              return;
          }

          dispatch_async(dispatch_get_main_queue(), ^{
            UITableViewCell *updateCell =
                (id)[tableView cellForRowAtIndexPath:indexPath];
            if (!updateCell) {
                return;
            }
            [(UIImageView *)[updateCell.contentView viewWithTag:13]
                setImage:image];
          });

        }];
    [task resume];

    return cell;
}
@end

File: EmarsysPredictShopSwift/EmarsysPredictShopSwift/ItemsManager.swift
//
//  ItemsManager.swift
//  EmarsysPredictShopSwift
//
//

import UIKit

class ItemsManager: NSObject, UITableViewDataSource, UITableViewDelegate {
    
    var items = [Item]()
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }
    
    func tableView(tableView: UITableView, numberOfSectionsInTableView section: Int) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("ItemCell")
        
        let item = items[indexPath.row]
        
        let title = cell?.contentView.viewWithTag(11) as! UILabel
        title.text = item.title
        
        let price = cell?.contentView.viewWithTag(12) as! UILabel
        price.text = String(format: "$%.02f", item.price!)
        
        let url = NSURL(string: item.image!)
        let session = NSURLSession.sharedSession()
        session.dataTaskWithURL(url!, completionHandler: {(data, response, error) in
            guard let data = data else { return }
            guard let image = UIImage(data: data) else { return }
            dispatch_async(dispatch_get_main_queue()) {
                guard let updateCell = tableView.cellForRowAtIndexPath(indexPath) else {
                    NSLog("Failed to get cell for index path %@", indexPath)
                    return
                }
                let imageView = updateCell.contentView.viewWithTag(13) as! UIImageView
                imageView.image = image
            }
        }).resume()
        
        return cell!
    }
}
